hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerProtocolServerSideTranslatorPB.java

```diff
@@ -215,6 +222,28 @@ private OMResponse internalProcessRequest(OMRequest request) throws
         //  Added the assertion.
         assert (omClientRequest != null);
         OMClientRequest finalOmClientRequest = omClientRequest;
+
+        ClientInvocationId clientInvocationId = ClientInvocationId.valueOf(
+            ClientId.valueOf(UUID.nameUUIDFromBytes(ProtobufRpcEngine.Server.getClientId())),
+            ProtobufRpcEngine.Server.getCallId());
+        RetryCache.Entry cacheEntry =
+            omRatisServer.getServerDivision().getRetryCache().getIfPresent(clientInvocationId);
+        // if this request has been executed before, return the cached result.
+        if (cacheEntry != null) {
+          if (cacheEntry.getReplyFuture().isDone()) {
+            RaftClientReply raftClientReply;
+            try {
+              raftClientReply = cacheEntry.getReplyFuture().get();
+            } catch (ExecutionException ex) {
+              throw new ServiceException(ex.getMessage(), ex);
+            } catch (InterruptedException ex) {
+              Thread.currentThread().interrupt();
+              throw new ServiceException(ex.getMessage(), ex);
+            }
+            return OMRatisHelper.getOMResponseFromRaftClientReply(raftClientReply);
+          }
+        }
```

Comment:
Shouldn't Ratis handle this kind of "replay"?

Reply:
No, it doesn't and that's what puzzled me initially. The replay happens after preExecute() in submitRequestToRatis().

Reply:
This method is executed by Hadoop RPC server handler (note that client sends requests to OM via Hadoop RPC, and the leader creates a raft request out of the Hadoop RPC request)

Reply:
Assuming it is needed and is specific to OM: please extract to a method to keep the level of abstraction in `internalProcessRequest` more consistent.

Something like:

```java
RaftClientReply raftClientReply = getCachedReply(...);
if (raftClientReply != null) {
  return OMRatisHelper.getOMResponseFromRaftClientReply(raftClientReply);
}
```

Reply:
Summary:

- Ratis does handle replay
- but replay in OM can fail in `preExecute` even before hitting Ratis
- so we workaround that by checking Ratis retry cache before `preExecute`